lib/SILOptimizer/IPO/GlobalOpt.cpp

```diff
@@ -770,6 +786,35 @@ void SILGlobalOpt::optimizeInitializer(SILFunction *AddrF,
   HasChanged = true;
 }
 
+/// Look through all the uses of the global variable. If the only use is a
+/// store, and the global is private, then we can remove the global, the store,
+/// and the alloc. Otherwise, return.
+void SILGlobalOpt::tryOptimizeUnusedGlobal(SILGlobalVariable *global,
+                                           StoreInst *store) {
+  if (global->getLinkage() != SILLinkage::Private)
+    return;
+  if (GlobalLoadMap.count(global) || GlobalAccessMap.count(global))
+    return;
+
+  auto globalAddr = cast<GlobalAddrInst>(store->getDest());
+  bool canRemove = true;
+  for (auto *use : globalAddr->getUses()) {
+    if (!isa<StoreInst>(use->getUser())) {
+      canRemove = false;
+      break;
+    }
+  }
+
+  if (canRemove) {
+    store->eraseFromParent();
```

Comment:
What if the global is a 'var' and there are multiple stores, but no loads? I'm wondering if it would be better to do this transformation in multiple phases:
- first find all globals that are never loaded
- then remove alloc_global and store instructions
- then remove the globals themselves

Reply:
I like your idea of multiple phases a lot. I updated this patch, as you suggested, and it is much cleaner/simpler. I have to iron out a few wrinkles; then, I'll push it.